macrosynergy/download/dataquery.py

```diff
@@ -0,0 +1,743 @@
+                crt=kwargs.pop("crt"),
+                key=kwargs.pop("key"),
+                base_url=kwargs.pop("base_url", CERT_BASE_URL),
+                proxy=self.proxy,
+            )
+
+        self.debug: bool = debug
+        self.last_url: Optional[str] = None
+        self.status_code: Optional[int] = None
+        self.last_response: Optional[str] = None
+        self.concurrent: bool = concurrent
+        self.batch_size: int = batch_size
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        if exc_type:
+            print(f"Execution {exc_type} with value (exc_value):\n{exc_value}")
+            logger.error(f"Execution {exc_type} with value (exc_value):\n{exc_value}")
+
+    def check_connection(self) -> Tuple[bool, dict]:
+        """Check connection (heartbeat) to DataQuery."""
+        endpoint = "/services/heartbeat"
+        js, success, msg = self.access.get_dq_api_result(
+            url=self.access.base_url + endpoint,
+            params={"data": "NO_REFERENCE_DATA"},
+            proxy=self.proxy,
+        )
+
+        if not success:
+            return False, {}
+
+        if "info" not in js:
+            logger.error(
+                f"Invalid response from DataQuery. {js}"
+                f"request {self.last_url:s} error response at {datetime.datetime.utcnow().isoformat()}: {js}"
+            )
+            raise ValueError(
+                f"Invalid response from DataQuery."
+                "'info' missing from response.keys():"
+                f"{js.keys()}, request {self.last_url:s} error response at {datetime.datetime.utcnow().isoformat()}: {js}"
+            )
+
+        results: dict = js["info"]
+        assert int(results["code"]) == 200, f"Error message from DataQuery: {results}"
+        return int(results["code"]) == 200, results
+
+    def _fetch_threading(self, endpoint, params: dict, server_count: int = 5):
+        """
+        Method responsible for requesting Tickers from the API. Able to pass in 20
+        Tickers in a single request. If there is a request failure, the function will
+        return a None-type Object and the request will be made again but with a slower
+        delay.
+
+        :param <str> endpoint:
+        :param <dict> params: dictionary containing the required parameters.
+        :param <int> server_count: count of servers to be retried.
+
+        return <dict>: singular dictionary obtaining maximum 20 elements.
+        """
+
+        # The url is instantiated on the ancillary Classes as it depends on the DQ access
+        # method chosen.
+        url = self.access.base_url + endpoint
+        select = "instruments"
+        response = {}
+        results = []
+        counter = 0
+        invalid_responses = 0
+        while (
+            (not (select in response.keys()))
+            and (counter <= server_count)
+            and (invalid_responses <= server_count)
```

Comment:
Wouldn't we arbitrarily break after 5 retries now as default? Even if the next link has more left?

Reply:
Yes; that's true. 
I think setting `server_count=10` and testing `(counter+invalid_responses)<=server_count` before breaking might be a better idea.
PS: the variables after refactor are 
`server_count`→`max_retries`
`counter`→`conxn_errors`